Dockerfile

```diff
@@ -10,17 +10,68 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
 USER root
 
-COPY marble_environment.yml /environment.yml
+ENV DEBIAN_FRONTEND noninteractive
+RUN apt-get update --yes && \
+    # - `apt-get upgrade` is run to patch known vulnerabilities in system packages
+    #   as the Ubuntu base image is rebuilt too seldom sometimes (less than once a month)
+    apt-get upgrade --yes && \
+    apt-get install --yes --no-install-recommends build-essential gfortran && \
+    apt-get clean && rm -rf /var/lib/apt/lists/*
 
-# Installing jupyter lab extensions in the main environment
-RUN mamba install --yes jupyterlab-git jupyter_conda jupyter-archive jupyter-server-proxy dask-labextension
-# Creating a "Marble" environment
-RUN mamba env create -f /environment.yml
+USER ${NB_UID}
+RUN set -x && \
+    # Installing jupyter lab extensions in the base environment
+    mamba install --yes jupyterlab-git=0.44.0 \
+    mamba_gator=5.2.1 \
+    jupytext=1.16.1 \
+    jupyter-archive=3.4.0 \
+    jupyter-server-proxy=4.1.0 \
+    dask-labextension=6.2.0 \
+    ipywidgets=8.1.2 \
+    jupyterlab=3.6.7 \
+    jupyter_bokeh=3.0.7 && \
+    mamba clean --all -f -y
 
-ENV PATH="/opt/conda/envs/marble/bin:$PATH"
+# Creating the "Marble" environment:
+COPY marble_environment.yml /tmp/environment.yml
+RUN set -x && \
+    mamba env create -f /tmp/environment.yml && \
+    mamba clean --all -f -y
 
-RUN python -m ipykernel install --name Marble
+# To get the marble version of pip for weaver installation:
+ENV PATH="$MAMBA_ROOT_PREFIX/envs/marble/bin:$PATH"
+# Installing weaver into the Marble environment:
+RUN set -x && \
+    git clone https://github.com/crim-ca/weaver.git && \
+    cd weaver && \
+    rm -rf .git && \
+    make install-pip
+
+# Installing s3fs separately because of issues with weaver dependencies
+RUN set -x && \
+    mamba install -n marble s3fs=2024.2.0 && \
+    mamba clean --all -f -y
+
+USER root
+# New kernels are installed in /usr/local/share/jupyter. Changing permissions of this location to be writable by the user.
+# This opens the possibility that in the jupyter environment a user can create a new environment and install a new kernel.
+RUN mkdir -p /usr/local/share/jupyter
+RUN fix-permissions /usr/local/share/jupyter
 
 USER ${NB_UID}
+# ENV PATH="$MAMBA_ROOT_PREFIX/envs/marble/bin:$PATH"
+RUN python -m ipykernel install --name Marble
 
-# CMD ["mamba", "run", "-n", "Marble", "/usr/local/bin/start-notebook.sh"]
\ No newline at end of file
+# For import xesmf since esmf-8.4.0, see
+# https://github.com/conda-forge/esmf-feedstock/issues/91
+ENV ESMFMKFILE=/opt/conda/envs/marble/lib/esmf.mk
+
+USER root
+# These need to be run as root
+RUN fix-permissions "/home/${NB_USER}/.ipython"
+RUN fix-permissions "/home/${NB_USER}/.cache"
+
+COPY .bashrc "/home/${NB_USER}/.bashrc"
```

Comment:
It works. Everything is correctly picked up. (and it should, the user is always jovyan and bash will read the `.bashrc` file in that user's directory). We'll have to test it when the container is launched with jupyterhub. To clarify, this is only to control the environment inside a terminal instance.

Reply:
Basically, this is what I am after in JupyterLab's terminal environment, and it works for that. Before this it was some odd `jovyan@bfa6f7052c2a` that was not very user friendly.
 
<img width="362" alt="Screenshot 2024-02-15 at 11 31 20 AM" src="https://github.com/DACCS-Climate/marble-jupyter-images/assets/5779123/db7d0267-af76-4702-8526-048f8de89ae2">

Reply:
Where does it work though? Are you testing it out with the full marble stack?

Reply:
Sorry, not sure I follow. I want the shell prompt changed in the JupyterLab environment. As I show in the screenshot, I can get that result.

Reply:
It's being launched as a stand-alone container by docker.

As I said earlier, we have to test this with jupyterHub (in the birdhouse-deploy). I am not 100% sure, but I don't think this should change if using JupyterHub. You think this is not going to work when deployed within birdhouse?

Reply:
Yeah. I don't think it will. But now you've got me second guessing myself so I'm testing it out on the test server

Reply:
I'll report back shortly

Reply:
Thanks! I'm not 100% sure myself, but let's make it interesting: loser buys beer. 

I think this time around you should be able to pull the amd64 image directly from docker hub.

Reply:
haha sure :)

Reply:
I'm just building the image from the dockerfile since I'm pretty sure the image on dockerhub isn't up to date with this branch right?

Reply:
it is this branch. I pushed it yesterday after the changes.

Reply:
![image](https://github.com/DACCS-Climate/marble-jupyter-images/assets/4380924/8f24a5c4-741d-41bf-bec8-e2149860fb96)

Reply:
:D 

Let me know whenever you want to claim the beer!

Alright.... not sure where this happened, but it must be in the DockerSpwaner stuff. Since you have this setup, can you test if copying to `/notebook_dir/writeable_workspace/.home` works?

Reply:
Unfortunately that's not going to work either because that whole directory is mounted as a volume so anything inside it when the container starts up is going to be overwritten.

I still think that the best place is to write it into /etc/profile.d . I would recommend doing: 

- remove the .bashrc file from this PR and add a new file that contains _only_ the additional bits you want to add (name it marble.sh or something)
- in the Dockerfile, add a line `COPY marble.sh /etc/profile.d/marble.sh` (run it as the root user)

Then we can test if that works on the test machine or if you want to test it locally you can deploy the container with:

```
docker run -it --rm -e HOME=/someotherdir testmarble bash
```

and you should be able to see right away if the PS1 changes worked for example

Reply:
So.... unfortunately this doesn't work exactly this way. Turns out `/etc/profile` is only sourced during a login shell, so putting anything in `/etc/profile` or `/etc/profile.d/*` only works when one does:
```
docker run -it --rm -e HOME=/someotherdir testmarble bash --login
```

However, `/etc/bash.bashrc` is [always sourced](https://askubuntu.com/a/40313), so adding the `PS1` command to `/etc/bash.bashrc` via:
```
RUN echo "PS1='\[\033[01;32m\]marbleUser\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\] >>> '" >> /etc/bash.bashrc
```
works when one starts a simple bash shell.

Note: when the `HOME` environment variable is not specified (to be something different than the default), then nothing works because the `PS1` command from `~/.bashrc` overrides that declared in any of the `/etc/*` files, but I think we don't need to care about this issue.